wgengine/router/ifconfig_windows.go

```diff
@@ -358,28 +362,39 @@ func configureInterface(cfg *Config, tun *tun.NativeTun) error {
 
 	ipif, err = iface.GetIpInterface(winipcfg.AF_INET6)
 	if err != nil {
-		return err
-	}
-	if err != nil && errAcc == nil {
-		errAcc = err
-	}
-	if foundDefault6 {
-		ipif.UseAutomaticMetric = false
-		ipif.Metric = 0
-	}
-	if mtu > 0 {
-		ipif.NlMtu = uint32(mtu)
-	}
-	ipif.DadTransmits = 0
-	ipif.RouterDiscoveryBehavior = winipcfg.RouterDiscoveryDisabled
-	err = ipif.Set()
-	if err != nil && errAcc == nil {
-		errAcc = err
+		if !isMissingIPv6Err(err) {
+			return err
+		}
+	} else {
+		if foundDefault6 {
+			ipif.UseAutomaticMetric = false
+			ipif.Metric = 0
+		}
+		if mtu > 0 {
+			ipif.NlMtu = uint32(mtu)
+		}
+		ipif.DadTransmits = 0
+		ipif.RouterDiscoveryBehavior = winipcfg.RouterDiscoveryDisabled
+		err = ipif.Set()
+		if err != nil && errAcc == nil {
+			errAcc = err
+		}
 	}
 
 	return errAcc
 }
 
+// isMissingIPv6Err reports whether err is due to IPv6 not being enabled on the machine.
+//
+// It's intended for use on errors returned by the winipcfg.Interface.GetIpInterface
+// method, which ultimately calls:
+// https://docs.microsoft.com/en-us/windows/win32/api/netioapi/nf-netioapi-getipinterfaceentry
+func isMissingIPv6Err(err error) bool {
+	// ERROR_NOT_FOUND from means the address family (IPv6) is not found.
+	// (ERROR_FILE_NOT_FOUND means that the interface doesn't exist.)
+	return errors.Is(err, windows.ERROR_NOT_FOUND)
```

Comment:
This error check is too general. ERROR_NOT_FOUND could be returned by many different Windows APIs in many different situations.

You use it to check winipcfg.GetIpInterface error. How do you know that you don't misinterpret some unrelated error as isMissingIPv6Err? We might be assuming that we don't have IPv6, while in reality our code just fails for a different reason. Do we care?

Also, I have never used error.Is, but, there is also syscall.ERROR_NOT_FOUND. How do you know that winipcfg.GetIpInterface returns windows.ERROR_NOT_FOUND and not windows.ERROR_NOT_FOUND? Does it matter?

Reply:
See the comment above it. It's only for use by errors returned by a certain method.

The actual error is an *os.SyscallError with its inner Err set to a syscall.Errno(windows.ERROR_NOT_FOUND).

I confirmed this works. You can right click network adapters and uncheck IPv6 and then it failed. With this change it no longer fails.

ERROR_NOT_FOUND means only address family not supported. (for this method)

Reply:
> You can right click network adapters and uncheck IPv6 and then it failed.

I am using code just before your change

```
commit cc1b2e6276b6d4bae473b5910f2c1671caf75d92 (HEAD)
Author: Brad Fitzpatrick <bradfitz@tailscale.com>
Date:   Fri Sep 11 12:49:46 2020 -0700

    cmd/tswin: fix LocalAppData

    The code adds the Tailscale later. It's not in the environment variable.

diff --git a/cmd/tswin/tswin.go b/cmd/tswin/tswin.go
index 9e239c67..40b9c519 100644
--- a/cmd/tswin/tswin.go
+++ b/cmd/tswin/tswin.go
@@ -151,7 +151,7 @@ func main() {
        be.SysProcAttr = &syscall.SysProcAttr{HideWindow: true}
        be.Stdout = os.Stdout
        be.Stderr = os.Stdout // merge the two for ease of redirection
-       be.Env = append(os.Environ(), `LocalAppData=C:\Windows\System32\config\systemprofile\AppData\Local\Tailscale`)
+       be.Env = append(os.Environ(), `LocalAppData=C:\Windows\System32\config\systemprofile\AppData\Local`)
        if err := be.Start(); err != nil {
                log.Fatalf("Starting backend: %v", err)
        }
```

I un-clicked IPv6 and pressed `OK`

![image](https://user-images.githubusercontent.com/9796621/92984779-25b26c80-f4f0-11ea-8690-bd841c55b247.png)

Then I run tailscale. I cannot find GetIpInterfaceEntry in my log (I should according to https://github.com/tailscale/tailscale/issues/412). And pinging 101.102.103.104 works fine too.

I also noticed that if I right click on Tailscale adapter, the IPv6 tick is back on.

![image](https://user-images.githubusercontent.com/9796621/92984837-b25d2a80-f4f0-11ea-9e5a-b365d08e18e3.png)

What am I doing wrong?

Reply:
Uncheck IPv6 on your other interface(s).

Reply:
> Uncheck IPv6 on your other interface(s).

I did, it did not help.

I have `Ethernet`, `Ethernet 3` and `Tailscale` network devices. I unticked IPv6 for them all. But when I start Tailscale `Tailscale` devices gets ticked again. And I don't see any error messages in the log.

Reply:
Are you on Windows 7? I was testing on Windows 10.

Reply:
I am using Windows 10 too

![image](https://user-images.githubusercontent.com/9796621/92985498-6319f880-f4f6-11ea-8ab5-1e4cf71f09a8.png)

Sorry I am using my `Snipping Tool` today.

Reply:
I just reproduced. I'm on Version 2004.

I un-checked IPv6 from both my normal ethernet & my Tailscale Tunnel, then

```patch
diff --git a/wgengine/router/ifconfig_windows.go b/wgengine/router/ifconfig_windows.go
index d6e101e42..08f6808dd 100644
--- a/wgengine/router/ifconfig_windows.go
+++ b/wgengine/router/ifconfig_windows.go
@@ -392,7 +392,10 @@ func configureInterface(cfg *Config, tun *tun.NativeTun) error {
 func isMissingIPv6Err(err error) bool {
        // ERROR_NOT_FOUND from means the address family (IPv6) is not found.
        // (ERROR_FILE_NOT_FOUND means that the interface doesn't exist.)
-       return errors.Is(err, windows.ERROR_NOT_FOUND)
+       if errors.Is(err, windows.ERROR_NOT_FOUND) {
+               panic(fmt.Sprintf("Got a %T: %#v, %v", err, err, err))
+       }
+       return false
 }
 
 // routeLess reports whether ri should sort before rj.
bradfitz@tsdev:~/src/tailscale.io/oss$ 
```

And ran Tailscale, to see that it does trigger:


```
pback Pseudo-Interface 1:true] HaveV6Global:false HaveV4:true IsExpensive:false DefaultRouteInterface:TODO HTTPProxy:http://10.0.0.52:3966}
srv: 1.6M/0.0M Routine: event worker - started
srv: 1.6M/0.0M router: dns: using dns.windowsManager
srv: 1.8M/0.0M UDP bind has been updated
WARNING: skipping windows socket binding.
panic: Got a *os.SyscallError: &os.SyscallError{Syscall:"iphlpapi.GetIpInterfaceEntry", Err:0x490}, iphlpapi.GetIpInterfaceEntry: Element not found.

goroutine 1 [running]:
tailscale.com/wgengine/router.isMissingIPv6Err(0x146f1a0, 0xc000088320, 0x0)
        tailscale.com@v0.0.0-00010101000000-000000000000/wgengine/router/ifconfig_windows.go:396 +0x139
tailscale.com/wgengine/router.monitorDefaultRoutes.func1(0x135e720, 0xc00004a460)
        tailscale.com@v0.0.0-00010101000000-000000000000/wgengine/router/ifconfig_windows.go:142 +0x388
tailscale.com/wgengine/router.monitorDefaultRoutes(0xc0001ae000, 0x1, 0xc0000488a0, 0xc00010fa78, 0x105c848, 0xc0001ae000)
        tailscale.com@v0.0.0-00010101000000-000000000000/wgengine/router/ifconfig_windows.go:159 +0x1ae
tailscale.com/wgengine/router.(*winRouter).Up(0xc000046b00, 0x0, 0x0)
        tailscale.com@v0.0.0-00010101000000-000000000000/wgengine/router/router_windows.go:58 +0x56
tailscale.com/wgengine.newUserspaceEngineAdvanced(0xc00003d520, 0x1480ae0, 0xc0000488a0, 0x13da5e0, 0x0, 0x0, 0x0, 0x0, 0x0)
        tailscale.com@v0.0.0-00010101000000-000000000000/wgengine/userspace.go:344 +0x915
tailscale.com/wgengine.NewUserspaceEngineAdvanced(...)
        tailscale.com@v0.0.0-00010101000000-000000000000/wgengine/userspace.go:198
tailscale.com/wgengine.NewUserspaceEngine(0xc00003d520, 0x13a86dd, 0x9, 0x13a0000, 0x11, 0x0, 0x0, 0xc00003d500)
        tailscale.com@v0.0.0-00010101000000-000000000000/wgengine/userspace.go:188 +0x2c5
tailscale.io/win/ui.startIPNServer.func1(0x13aef4a, 0x11, 0x0, 0x0)
        tailscale.io/win/ui/main.go:35 +0x51
tailscale.io/win/ui.startIPNServer(0x13a6911, 0x5, 0x147b680, 0xc0000180b8, 0xc00001c1c0, 0x40, 0xc000160690, 0xc00010fec8)
        tailscale.io/win/ui/main.go:56 +0x1bc
tailscale.io/win/ui.Main()
        tailscale.io/win/ui/main.go:123 +0xced
main.main()
        tailscale.io/win/wg64/main.go:13 +0x27
subprocess exited: exit status 2
```

And if I instead do:

```patch
diff --git a/wgengine/router/ifconfig_windows.go b/wgengine/router/ifconfig_windows.go
index d6e101e42..4c740ac8d 100644
--- a/wgengine/router/ifconfig_windows.go
+++ b/wgengine/router/ifconfig_windows.go
@@ -390,9 +390,7 @@ func configureInterface(cfg *Config, tun *tun.NativeTun) error {
 // method, which ultimately calls:
 // https://docs.microsoft.com/en-us/windows/win32/api/netioapi/nf-netioapi-getipinterfaceentry
 func isMissingIPv6Err(err error) bool {
-       // ERROR_NOT_FOUND from means the address family (IPv6) is not found.
-       // (ERROR_FILE_NOT_FOUND means that the interface doesn't exist.)
-       return errors.Is(err, windows.ERROR_NOT_FOUND)
+       return false
 }

 // routeLess reports whether ri should sort before rj.
```

Then I get:

```
srv: 1.8M/0.0M Routine: event worker - started
srv: 1.9M/0.0M router: dns: using dns.windowsManager
srv: 2.0M/0.0M UDP bind has been updated
WARNING: skipping windows socket binding.
MonitorDefaultRoutes: iphlpapi.GetIpInterfaceEntry: Element not found.
subprocess exited: exit status 1
```

So the change definitely works for me.

Reply:
I sent https://github.com/tailscale/tailscale/pull/758 to tighten the error check.

Reply:
> I just reproduced. I'm on Version 2004.

Yes my version is behind. Not sure if that matters.

> I un-checked IPv6 from both my normal ethernet & my Tailscale Tunnel

Does your Tailscale Tunnel IPv6 stays unticked after you run tailscale? Because my gets tick on automatically.

> then
> 
> ```diff
> diff --git a/wgengine/router/ifconfig_windows.go b/wgengine/router/ifconfig_windows.go
> index d6e101e42..08f6808dd 100644
> --- a/wgengine/router/ifconfig_windows.go
> +++ b/wgengine/router/ifconfig_windows.go
> @@ -392,7 +392,10 @@ func configureInterface(cfg *Config, tun *tun.NativeTun) error {
>  func isMissingIPv6Err(err error) bool {
>         // ERROR_NOT_FOUND from means the address family (IPv6) is not found.
>         // (ERROR_FILE_NOT_FOUND means that the interface doesn't exist.)
> -       return errors.Is(err, windows.ERROR_NOT_FOUND)
> +       if errors.Is(err, windows.ERROR_NOT_FOUND) {
> +               panic(fmt.Sprintf("Got a %T: %#v, %v", err, err, err))
> +       }
> +       return false
>  }
>  
>  // routeLess reports whether ri should sort before rj.
> bradfitz@tsdev:~/src/tailscale.io/oss$ 
> ```
> 
> And ran Tailscale, to see that it does trigger:

Mine does not panic.

I am not sure what else we could try.

Reply:
Did you stop Tailscale before unchecking them?

In any case, you don't need to continue to reproduce it. I'm confident in the fix.

And yes, we recreate the interface at start-up. But on my system, the re-created interface retains its IPv6 state (stays checked or stays unchecked). Not sure why.

Reply:
> Did you stop Tailscale before unchecking them?

I did. I just tried that again.

> In any case, you don't need to continue to reproduce it. I'm confident in the fix.

Fine with me. I will stop debuging this.